pywbem/cim_operations.py

```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
+              causes the server-implemented default behaviour to be used.
+              :term:`DSP0200` defines that the server-implemented default is
+              to return zero instances.
+
+        Keyword Arguments:
+
+          extra :
+            Additional keyword arguments are passed as additional operation
+            parameters to the WBEM server.
+            Note that :term:`DSP0200` does not define any additional parameters
+            for this operation.
+
+        Raises:
+
+            Exceptions described in :class:`~pywbem.WBEMConnection`.
+
+        Returns:
+
+          A Python iterator object. Instances can be retrieved
+          by iterating through the iterator.
+
+        Example:
+            enum_gen = conn.EnumerateInstancesWithPathGen('CIM_Blah')
+            for inst in enum_gen:
+                if inst.get('thisPropertyName') == 0
+                    enum_get.close()
+                    break;
+                else:
+                    print('instance %s' % inst.tomof())
+
+        """
+
+        if MaxObjectCount is not None and MaxObjectCount <= 0:
+            raise ValueError('MaxObjectCount=%s. Invalid" le 0' % \
+                             MaxObjectCount)
+
+        # common variable for pull result tuple so it can be used by finally:
+        pull_result = None
+        try:                # try / finally block to allow close()
+            if self.has_pull_operations is None or self.has_pull_operations:
+                try:
+                    pull_result = self.OpenEnumerateInstances(
+                        ClassName, namespace=namespace, LocalOnly=LocalOnly,
+                        DeepInheritance=DeepInheritance,
+                        IncludeQualifiers=IncludeQualifiers,
+                        IncludeClassOrigin=IncludeClassOrigin,
+                        PropertyList=PropertyList,
+                        FilterQueryLanguage=FilterQueryLanguage,
+                        FilterQuery=FilterQuery,
+                        OperationTimeout=OperationTimeout,
+                        ContinueOnError=ContinueOnError,
+                        MaxObjectCount=MaxObjectCount, **extra)
+
+                    for inst in pull_result.instances:
+                        yield inst
+
+                    # loop to pull while more exist.
+                    while not pull_result.eos:
+                        pull_result = self.PullInstancesWithPath(
+                            pull_result.context, MaxObjectCount=MaxObjectCount)
+
+                        while inst in pull_result.instances:
+                            yield inst
+                    return
+
+                # if Not supported, set flag and try alternative operation.
+                except CIMError as ce:
+                    if ce.args[0] != CIM_ERR_NOT_SUPPORTED:
+                        self.has_pull_operations = False
+                    else:
+                        raise
```

Comment:
Here is a logic error. The description says we want to try pull enumerations if `has_pull_operations` is None, and set it to True if that succeeds. That second part does not happen in the current code, it can be done in an else path of the try/except.